Fix ESM, CommonJS and script building

asrashley opened the pull request:
Multiple failures have been discovered in the build rules that generate the three different build flavours. See #5 

This PR attempts to fix these issues and add tests to try to stop them going wrong in the future.

Commit:
add vitest package

Commit:
exports should point to built code not source

The exports section should reference the built code, not the original
source. This is required so that the correct import syntax is used for
each varient.

Commit:
use vitest to run tests

Commit:
browser and node.js use different mounting for the library

the webpack config files for ESM and CJS need different build flags.

CommonJS modules should be using "this" rather than "self" to install
the library. Without that change, the require() call fails with
"self is undefined" error.

When using the ESM module within a browser, "self" is used during
the import of the module.

Commit:
add test of using build after creation

once the build has been created, run a unit test to check that the
built library can be correctly imported.

Commit:
rename HTML template

Rename the HTML template to use a name that makes it clear that it
is an HTML file rather than a JavaScript file.

Commit:
use optimised build

Commit:
fix webpack commonjs output to generate CJS format

Commit:
fix outputHTML used before initialisation

Commit:
fix webpack script output to generate browser script format

Commit:
add tests of using builds within a browser

Add two examples of using the codec-string library as a either a
simple <script> tag that adds the library to the global scope, or
as an ESM module.

Commit:
update examples to match current API

Commit:
add documentation of running tests

paulhiggs commented:
The test process is excellent, thanks.
While the build for `esm` and `script` modes looks OK, the `dist\cjs\index.html` does not include/import `app.js` or `codec-string.js`, i.e...
````
<!doctype html><html>
<head><title>Codec string decoder (parse+print mode)</title></head>
<body>
<h1>Codec string decoder (parse+print mode)</h1>
<p>codec-string v0.1.6 by <a href="mailto:paul_higgs@hotmail.com">Paul Higgs</a> ( <a href="https://github.com/paulhiggs/codec-string">repository</a>, <a href="https://github.com/paulhiggs/codec-string/issues">issues</a> )</p><p>Enter string: codecs="<input id="param"/>" <button type="button" id="check-codec">Decode</button></p><div id="results"/>
</body></html>
````

Commit:
upgrade Node.js version to v20

Current LTS is v22, but this upgrade is sufficient to meet
the engine requirements of the packages in package.json

asrashley commented:
> The test process is excellent, thanks. While the build for `esm` and `script` modes looks OK, the `dist\cjs\index.html` does not include/import `app.js` or `codec-string.js`, i.e...

This is because CommonJS is not suitable for use in the browser. It's only for use in Node.js. It shouldn't really be generating an index.html file for the cjs build. An example JS app would be more useful.

https://aaronbos.dev/posts/commonjs-vs-es-module

We could add yet another flavour `AMD` that can be used in the browser.

Commit:
add del-webpack-plugin

Commit:
delete index.html from CommonJS build

As the CommonJS build is only meant for usage with Node.js, having
an index.html file in its output is confusing.

Commit:
move ESM import example into examples directory

To make it easier to find the example usage code, put the JS code
into the /examples directory, rather than embedding it directly
into the HTML test file.

Commit:
add copy-webpack-plugin

Commit:
copy an example Node.js app into the dist/cjs directory

After the CommonJS build has completed, copy an example cjs script
into the output. This provides a handy example and a way to check
from the command line if the build is working in Node.js

asrashley commented:
I have updated the CommonJS build to remove the HTML files and instead copy a Node.js example.

```
dist/cjs:
total 104
-rw-rw-r-- 1 alex alex 99617 Jun  6 12:09 codec-string.cjs
-rw-rw-r-- 1 alex alex   122 Jun  6 12:09 node_app.cjs
```

The `node_app.cjs` runs a simple test of a codec decode.

```
$ (cd dist/cjs/ && node ./node_app.cjs )
e {
  decodes: [
    e {
      label: 'AVC/H.264',
      parsed: [Array],
      HTMLPrinter: [Function: u]
    }
  ]
}
<span style="font-weight:bold">AVC/H.264</span><br>profile_idc=100 constraint_set=0 level_idc=42<br>profile=High (64)<br>constraints=&#8209;&#8209;&#8209;&#8209;&#8209;&#8209;<br>level=4.2 (2a)<br><br><span style="font-weight:bold">DVB term: </span>urn:dvb:metadata:cs:VideoCodecCS:2022:1.4.14<br>
```
